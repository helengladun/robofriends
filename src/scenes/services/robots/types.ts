export enum RobotsTypes {
  REQUEST_ROBOTS = 'REQUEST_ROBOTS',
  REQUEST_ROBOTS_SUCCESS = 'REQUEST_ROBOTS_SUCCESS',
  REQUEST_ROBOTS_FAILED = 'REQUEST_ROBOTS_FAILED'
}