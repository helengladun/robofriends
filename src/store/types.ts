export enum SearchTypes {
  CHANGE_SEARCH_FIELD = 'CHANGE_SEARCH_FIELD'
}