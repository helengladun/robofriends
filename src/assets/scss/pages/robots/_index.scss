@import 'main-title';
@import 'scroll';