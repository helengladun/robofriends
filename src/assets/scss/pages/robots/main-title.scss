.main-title {
  color: var(--primary-title-color);
  font-family: var(--primary-headers-font);
  font-weight: 200;
}