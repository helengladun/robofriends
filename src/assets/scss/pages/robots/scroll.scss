.scroll {
  height: 500px;
}