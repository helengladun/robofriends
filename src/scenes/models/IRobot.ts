import {IAddress} from './IAddress';
import {ICompany} from './ICompany';

export interface IRobot {
  id: number,
  name: string,
  username?: string,
  email: string,
  address?: IAddress,
  phone?: string,
  website?: string,
  company?: ICompany
}