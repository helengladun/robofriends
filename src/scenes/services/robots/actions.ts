// types
import { RobotsTypes } from './types';

// interfaces
import { IRobot } from '../../models/IRobot';

export const requestRobots = () => ({
  type: RobotsTypes.REQUEST_ROBOTS
});

export const requestRobotsSuccess = (data: IRobot[]) => ({
  type: RobotsTypes.REQUEST_ROBOTS_SUCCESS,
  payload: {
    data
  }
});

export const requestRobotsFailed = (error: string) => ({
  type: RobotsTypes.REQUEST_ROBOTS_FAILED,
  payload: {
    error
  }
});
