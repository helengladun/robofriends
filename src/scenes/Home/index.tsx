import React, { useCallback, useEffect, useState } from 'react';
import debounce from 'lodash/debounce';
import { useDispatch, useSelector } from 'react-redux';

// components
import Spinner from '../components/Spinner';
import Header from '../components/Header';
import SearchBox from './components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundary from '../components/ErrorBoundary';
import CardList from './components/CardList';

// interfaces
import { IRobot } from '../models/IRobot';

// actions
import { requestRobots } from '../services/robots/actions';
import { setSearchField } from '../services/search/actions';

// reducers
import { IApplicationState } from '../../store/rootReducer';
import { IRobotsState } from '../services/robots/reducers';
import { ISearchState } from '../services/search/reducers';

const App = () => {
    const dispatch = useDispatch();
  const { robotsList, error, loading }: IRobotsState = useSelector(
    (state: IApplicationState) => state.robots
  );
  const { searchField }: ISearchState = useSelector((state: IApplicationState) => state.search);
  const [searchResult, setSearchResults] = useState<IRobot[]>([]);

  const debouncedChange = debounce(value => {
    dispatch(setSearchField(value));
  }, 500);

  const onChangeHandler = useCallback(
    (event: React.SyntheticEvent<HTMLInputElement>): void => {
      event.persist();
      debouncedChange(event.currentTarget.value);
    },
    [debouncedChange]
  );

  useEffect(() => {
    dispatch(requestRobots());
  }, [dispatch]);

  useEffect(() => {
    const results = robotsList.filter(
      (robot: IRobot) => robot.name.toLowerCase().search(searchField.toLowerCase()) > -1
    );
    setSearchResults(results);
  }, [robotsList, searchField]);

  if (loading) {
    return <Spinner />;
  }

  return (
    <div className="tc">
      <Header />
      <SearchBox onChangeHandler={onChangeHandler} />
      {error && <div className="red">{error}</div>}
      {robotsList && robotsList.length > 0 && (
        <Scroll>
          <ErrorBoundary>
            {searchField === '' ? (
              <CardList robots={robotsList} />
            ) : (
              <CardList robots={searchResult} />
            )}
          </ErrorBoundary>
        </Scroll>
      )}
    </div>
  );
};

export default App;
