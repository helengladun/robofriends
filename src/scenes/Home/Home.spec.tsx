import React from 'react';
import { shallow, mount } from 'enzyme';
import {useSelector} from 'react-redux';

import Home from './index';

import {IRobot} from '../models/IRobot';

import {RobotsTypes} from '../services/robots/types';

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
  useSelector: jest.fn()
}));

const robotsMock = [
  {
    id: 1,
    name: 'user1',
    email: 'user1@gmail.com'
  },
  {
    id: 2,
    name: 'test1',
    email: 'test@gmail.com'
  }
];

describe('App component', () => {

  it('render App component correctly', () => {
    (useSelector as jest.Mock<{robotsList: IRobot[]}>).mockImplementation(() => ({
      robotsList: robotsMock
    }));
    const component = shallow(<Home />);
    expect(component).toMatchSnapshot();
  });

  it('do not render list of robots', () => {
    (useSelector as jest.Mock<{robotsList: IRobot[]}>).mockImplementation(() => ({
      robotsList: []
    }));
    const component = shallow(<Home />);
    expect(component.find('Memo(Spinner)').exists()).toBeFalsy();
    expect(component.find('Scroll').exists()).toBeFalsy();
    expect(component.find('ErrorBoundary').exists()).toBeFalsy();
    expect(component.find('Memo(CardList)').exists()).toBeFalsy();
    expect(component.find('div.red').exists()).toBeFalsy();
  });

  it('render list of robots', () => {
    (useSelector as jest.Mock<{robotsList: IRobot[]}>).mockImplementation(() => ({
      robotsList: robotsMock
    }));
    const component = shallow(<Home />);
    expect(component.find('Scroll').exists()).toBeTruthy();
    expect(component.find('ErrorBoundary').exists()).toBeTruthy();
    expect(component.find('Memo(CardList)').exists()).toBeTruthy();
  });

  it('useEffect in App', () => {
    (useSelector as jest.Mock<{robotsList: IRobot[]; searchField: string}>).mockImplementation(() => ({
      robotsList: robotsMock,
      searchField: 'use'
    }));

    const component = mount(<Home />);

    expect(mockDispatch).toHaveBeenCalledWith({
      type: RobotsTypes.REQUEST_ROBOTS
    });

    expect(component.find('Memo(CardList)').prop('robots')).toEqual([{...robotsMock[0]}]);
  });

  it('render error block', () => {
    (useSelector as jest.Mock<{error: string}>).mockImplementation(() => ({
      error: 'error'
    }));

    const component = shallow(<Home />);
    expect(component.find('div.red').exists()).toBeTruthy();
  });
});