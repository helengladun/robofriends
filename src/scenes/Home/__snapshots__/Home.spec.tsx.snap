// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component render App component correctly 1`] = `
<div
  className="tc"
>
  <Memo(Header) />
  <SearchBox
    onChangeHandler={[Function]}
  />
  <Scroll>
    <ErrorBoundary>
      <Memo(CardList)
        robots={Array []}
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
